### PR TITLE
[SPARK-33551][SQL] Do not use custom shuffle reader for repartition

### PR DESCRIPTION
**Pull request by maryannxue**
### What changes were proposed in this pull request?

This PR fixes an AQE issue where local shuffle reader, partition coalescing, or skew join optimization can be mistakenly applied to a shuffle introduced by repartition or a regular shuffle that logically replaces a repartition shuffle.
The proposed solution checks for the presence of any repartition shuffle and filters out not applicable optimization rules for the final stage in an AQE plan.

### Why are the changes needed?

Without the change, the output of a repartition query may not be correct.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added UT.

### CODE REVIEW

**Comment by maryannxue** [comment]
cc @cloud-fan @manuzhang

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 1]
```diff
@@ -55,6 +55,11 @@ import org.apache.spark.sql.internal.SQLConf
  */
 object OptimizeSkewedJoin extends Rule[SparkPlan] {
 
+  /**
+   * The list of [[ShuffleOrigin]]s supported by this rule.
+   */
+  val supportedShuffleOrigins: Seq[ShuffleOrigin] = Seq(ENSURE_REQUIREMENTS)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:61]: should the rule `OptimizeSkewedJoin` itself check it?

**Comment by cloud-fan** [comment]
It's unfortunate that we can't retain the shuffle origin info when we optimize out the repartition shuffle. The approach here looks like a good compromise.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala (hunk 1 above)
Comment (maryannxue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:61]: Yes, we better do, to make it more robust, although it's not really necessary here as we exclude these rules if there's been a repartition shuffle.

**Review by cloud-fan — APPROVED** [review]

**Comment by gatorsmile** [comment]
Thanks! Merged to master

**Comment by manuzhang** [comment]
@maryannxue, @cloud-fan, 

Sorry for not raising up earlier but I'd like to discuss a case which seems to not have been covered here.

1. The final stage *before write* is a `SortMergeJoin` with partitioning that match the target table.
2. AQE switches the `SortMergeJoin` to `BroadcastHashJoin` as one side is smaller than broadcast threshold.
3. The probe side, with a different partitioning, is applied `OptimizeLocalShuffleReader`, which breaks the user intended partitioning.

**Comment by cloud-fan** [comment]
is `repartition` involved in your case?

**Comment by maryannxue** [comment]
I think the case you mentioned is exactly one of the test cases. Can you
find a case that is not working? And if repartition is not specified, we
should not guarantee any partitioning.

**Comment by manuzhang** [comment]
@cloud-fan 
No explicit `repartition` but implicit through join with `spark.sql.shuffle.partitions` configuration.

**Comment by cloud-fan** [comment]
then it's not a user-specified partitioning and we don't need to respect it.

**Comment by manuzhang** [comment]
> then it's not a user-specified partitioning and we don't need to respect it.

why ? User set  `spark.sql.shuffle.partitions`

**Comment by cloud-fan** [comment]
> why ? User set spark.sql.shuffle.partitions

Then we can't do AQE at all as it changes the number of partitions...

`spark.sql.shuffle.partitions` is a config for internal shuffles, and it does not have any guarantee for user-visible behaviors.

**Comment by manuzhang** [comment]
If it's not the final stage, `OptimizeLocalShuffleReader` will not be applied to the probe side when the output partitioning can not `ensureRequirements`. I think we lack this `ensureRequirements` for the final stage, which is usually the requirement of target table.

**Comment by cloud-fan** [comment]
The final stage can be just `df.collect()` where there is no target table. To be clear, we can't make any guarantee about the internal shuffles, as it's not reliable (depends on join strategy, aggregate strategy, etc.). We only need to respect user-given `repartition`.

**Comment by manuzhang** [comment]
> To be clear, we can't make any guarantee about the internal shuffles, as it's not reliable (depends on join strategy, aggregate strategy, etc.)

When it's the final stage, it's crossing the boundary and interacting with target table and downstream jobs. It's no longer internal and I think we need have some guarantee here.

**Comment by maryannxue** [comment]
The output partitioning and sort order are totally implementation
dependent. Say an SMJ would preserve partitioning from both sides while a
BHJ would only do one side. And we could well have any random join
implementation that does not preserve partititioning from either side. And
how can spark possibly guarantee that?

On the other hand, if the user asks for a certain partitioning, they should
specify it by using "repartition", and it's spark's job to optimize it out
if possible.

**Comment by manuzhang** [comment]
@maryannxue, @cloud-fan 
I got your point. I just wish we could do more to create less surprise for users.
Anyway, thanks for patiently answering my questions through holidays.

**Pull request closed without merging**